lib/model/rwfolder.go

```diff
@@ -760,33 +770,70 @@ func (f *sendReceiveFolder) deleteDir(file protocol.FileInfo, matcher *ignore.Ma
 		})
 	}()
 
-	// Delete any temporary files lying around in the directory
-
 	files, _ := f.fs.DirNames(file.Name)
+	// Detect content in the dir that required the dir itself to be
+	// resurrected, schedule unknown content to be scanned and find temporary
+	// and deletable ignored files.
+	resurrect := false
+	hasIgnored := false
+	hasToBeScanned := false
+	toBeDeleted := make([]string, 0)
 	for _, dirFile := range files {
 		fullDirFile := filepath.Join(file.Name, dirFile)
 		if fs.IsTemporary(dirFile) || (matcher != nil && matcher.Match(fullDirFile).IsDeletable()) {
-			f.fs.RemoveAll(fullDirFile)
+			toBeDeleted = append(toBeDeleted, fullDirFile)
+		} else if matcher != nil && matcher.Match(fullDirFile).IsIgnored() {
+			hasIgnored = true
+		} else if cf, ok := f.model.CurrentFolderFile(f.ID, fullDirFile); ok && !cf.IsDeleted() && !cf.IsInvalid() {
+			resurrect = true
+		} else {
+			// Something appeared in the dir that we either are not
+			// aware of at all, that we think should be deleted or that
+			// is invalid, but not currently ignored -> schedule scan
+			toBeScanned[fullDirFile] = struct{}{}
+			hasToBeScanned = true
 		}
 	}
 
+	if resurrect {
+		l.Debugln(f, "resurrecting dir", file.Name)
+		// A new file has been added before the dir was deleted -> revive the dir in index
+		file.Version = file.Version.Update(f.model.shortID)
+		dbUpdateChan <- dbUpdateJob{file, dbUpdateResurrectDir}
+		return
+	}
```

Comment:
Isn't it easier on each dbHandle/ShortcutFile to traverse up each directory up to the root and bump the version?

The rest of checks and errors etc look good.

Reply:
maybe easier but very expensive: on every file change you need as many additional database lookups as there is path components.

Reply:
As, isn't it easier to handle this at the scanner.

Reply:
Yeah, but technically that's the right thing to do to prevent races. This seems like a hack.
Furthermore, if we go for a graph based puller, doing this makes perfect sense?

Reply:
In my opinion this is a pretty clean hack and preferable to the performance hit of updating all parent folders on every detected change via pull or scan, but that's obviously a judgement call.
As for the graph based puller: I don't know what it is and how it relates to this, but I hear about it every now and then whenever something complex comes up - this sounds like one of these future wonders to be completed yesterday things. More to the point: I don't think it is a good idea to postpone something because of it.

Reply:
I'll leave for @calmh to make the judgement call. This just feels like technical debt that potentially will never be repaid.

Reply:
While fixing https://github.com/syncthing/syncthing/pull/4493#discussion_r149401957 I thought of another very unlikely but possible corner case which would make this entire thing even more complex/ugly. I am getting increasingly convinced that this is not the way to go. You might want to hold off reviewing/judgment calling this PR for now, I'll do some more digging/thinking into the issue itself.